### PR TITLE
DataCollectionMonitor: redesign config file

### PR DESCRIPTION
**Pull request by vlin17**
In the past, we directly defined categories in the config file, which results ~5000 lines of repetitive information. This change alleviates the pain by automatically generating categories for each scenario from the given features (e.g. speed) and their ranges (e.g. 0 ~ 10 mps, 10 ~ 30 mps) in the new config format. 

Note: categories is all possible combination of features x ranges.

### CODE REVIEW

File: modules/dreamview/backend/data_collection_monitor/data_collection_monitor.h [hunk 1]
```diff
@@ -90,6 +95,10 @@ class DataCollectionMonitor {
   // The table defines data collection requirements for calibration
   DataCollectionTable data_collection_table_;
 
+  // A map from scenario to its categories. Categories are collections
+  // of criteria from all possible combination of Feature x Range in a Scenario.
+  std::unordered_map<std::string, std::vector<Range>> categories_;
+
```
Comment (wanglei828) [inline modules/dreamview/backend/data_collection_monitor/data_collection_monitor.h]: Rename "categories_" for clarity

File: modules/dreamview/backend/data_collection_monitor/data_collection_monitor.h [hunk 2]
```diff
@@ -77,10 +78,14 @@ class DataCollectionMonitor {
  private:
   void InitReaders();
   void LoadConfiguration();
+  void ConstructCategories();
+  void ConstructCategoriesHelper(const Scenario& scenario, int feature_idx,
+                                 const Range& current_category,
+                                 std::vector<Range>* categories);
```
Comment (wanglei828) [inline modules/dreamview/backend/data_collection_monitor/data_collection_monitor.h]: "categories" should be vector<vector<Range>>.

File: modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc [hunk 3]
```diff
@@ -124,23 +125,66 @@ void DataCollectionMonitor::LoadConfiguration() {
...
+    Category new_category(current_category);
+    new_category.push_back(range);
+
+    // set new category name by appending it with the current range name
+    std::stringstream new_category_name;
+    new_category_name << current_category_name;
+    const std::string& range_name = range.name();
+    if (feature.range().size() > 1 && !range_name.empty()) {
+      if (!current_category_name.empty()) {
+        new_category_name << ", ";
```
Comment (unacao) [inline modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc]: I think we can use += to append to the category name, should be more efficient than <<

File: modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc [hunk 4]
```diff
@@ -124,23 +125,66 @@ void DataCollectionMonitor::LoadConfiguration() {
...
+void DataCollectionMonitor::ConstructCategories() {
+  scenario_to_categories_.clear();
+
+  for (const auto& scenario_iter : data_collection_table_.scenario()) {
+    const std::string& scenario_name = scenario_iter.first;
+    const Scenario& scenario = scenario_iter.second;
+
+    Category category;
+    ConstructCategoriesHelper(scenario, 0, "", category,
+                              &scenario_to_categories_[scenario_name]);
```
Comment (unacao) [inline modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc]: maybe pass in the scenario_name and let the helper access scenario_to_categories_ directly?

File: modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc [hunk 5]
```diff
@@ -124,23 +125,66 @@ void DataCollectionMonitor::LoadConfiguration() {
...
 
-    for (const auto& category_iter : scenario.category()) {
+    for (const auto& category_iter : categories) {
       const std::string& category_name = category_iter.first;
-      const Category& category = category_iter.second;
-
       category_consecutive_frame_count_[scenario_name][category_name] = 0.0;
       category_frame_count_[scenario_name][category_name] = 0.0;
-      current_progress_json_[scenario_name][category.description()] = 0.0;
+      current_progress_json_[scenario_name][category_name] = 0.0;
```
Comment (unacao) [inline modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc]: can we move these 3 lines of initialization to the end condition of the recursion? i.e. after line 162

**Pull request merged**